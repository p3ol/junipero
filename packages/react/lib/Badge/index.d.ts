import {
  ReactNode,
  ElementType,
  MutableRefObject,
  ComponentPropsWithRef,
} from 'react';

export declare type BadgeRef = {
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

declare interface BadgeProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  tag?: string | ElementType;
  ref?: MutableRefObject<BadgeRef | undefined>;
}

declare function Badge(props: BadgeProps): ReactNode | JSX.Element;

export default Badge;
