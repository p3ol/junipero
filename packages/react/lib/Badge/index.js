import { forwardRef, useImperativeHandle, useRef } from 'react';
import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

const Badge = forwardRef(({
  className,
  tag: Tag = 'span',
  ...rest
}, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    isJunipero: true,
    innerRef,
  }));

  return (
    <Tag
      { ...rest }
      ref={innerRef}
      className={classNames('junipero', 'badge', className)}
    />
  );
});

Badge.displayName = 'Badge';
Badge.propTypes = {
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
  ]),
};

export default Badge;
