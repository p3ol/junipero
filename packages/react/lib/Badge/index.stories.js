import Badge from './index';

export default { title: 'junipero/Badge' };

export const basic = () => (
  <>
    <p><Badge>1</Badge></p>
    <p><Badge className="primary">12</Badge></p>
    <p><Badge className="secondary">1</Badge></p>
    <p><Badge className="ternary">1</Badge></p>
    <p><Badge className="warning">1</Badge></p>
    <p><Badge className="danger">1</Badge></p>
    <p><Badge className="success">1</Badge></p>
  </>
);
