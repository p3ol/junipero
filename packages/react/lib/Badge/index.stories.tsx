import { useRef } from 'react';

import Badge, { BadgeRef } from './index';

export default { title: 'react/Badge' };

export const basic = () => {
  const ref = useRef<BadgeRef>();
  const isJunipero: boolean = ref.current?.isJunipero ?? false;

  return (
    <>
      <Badge ref={ref}>10</Badge>
      <Badge className="primary">1</Badge>
      <Badge className="danger">1</Badge>
      <Badge className="warning">1</Badge>
      <Badge className="success">1</Badge>
    </>
  );
};

export const smaller = () => (
  <>
    <Badge className="info">1</Badge>
    <Badge className="info primary">1</Badge>
    <Badge className="info danger">1</Badge>
    <Badge className="info warning">1</Badge>
    <Badge className="info success">1</Badge>
  </>
);
