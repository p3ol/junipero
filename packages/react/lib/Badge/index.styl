@require "../theme/colors"

.junipero.badge
  --main-color: var(--default-main-color)
  --sub-color: var(--default-sub-color)

  background: var(--main-color)
  color: var(--sub-color)
  border-radius: 10px
  font-weight: 700
  display: inline-block
  font-size: 10px
  text-align: center
  height: 20px
  line-height: 20px
  min-width: 20px
  padding: 0 4px

  for type in (primary secondary ternary warning danger success)
    &.{type}
      for color in (main sub)
        {'--' + color}-color: "var(--%s-%s-color)" % (type color)
