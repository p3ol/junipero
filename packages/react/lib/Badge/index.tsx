import {
  ComponentPropsWithRef,
  ElementType,
  MutableRefObject,
  ReactNode,
  forwardRef,
  useImperativeHandle,
  useRef,
} from 'react';
import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

import { ForwardedProps } from '../utils';

export declare type BadgeRef = {
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface BadgeProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  tag?: string | ElementType;
  ref?: MutableRefObject<BadgeRef>;
}

const Badge = forwardRef(({
  className,
  tag: Tag = 'span',
  ...rest
}: BadgeProps, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    isJunipero: true,
    innerRef,
  }));

  return (
    <Tag
      { ...rest }
      ref={innerRef}
      className={classNames('junipero', 'badge', className)}
    />
  );
}) as ForwardedProps<BadgeProps, BadgeRef>;

Badge.displayName = 'Badge';
Badge.propTypes = {
  tag: PropTypes.any, //TODO fixme
};
export default Badge;
