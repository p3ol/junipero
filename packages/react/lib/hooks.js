import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
