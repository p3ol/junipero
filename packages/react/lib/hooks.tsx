import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
