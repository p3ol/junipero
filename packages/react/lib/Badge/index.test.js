import { createRef } from 'react';
import { render } from '@testing-library/react';

import Badge from './index';

describe('<Badge />', () => {
  it('should render', () => {
    const ref = createRef();
    const { unmount, container } = render(<Badge ref={ref}>1</Badge>);
    expect(container).toMatchSnapshot();
    expect(ref.current.isJunipero).toBe(true);
    expect(ref.current.innerRef.current).toBe(container.firstChild);
    unmount();
  });
});
