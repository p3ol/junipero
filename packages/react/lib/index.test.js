import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
