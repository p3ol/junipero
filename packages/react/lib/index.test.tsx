import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
