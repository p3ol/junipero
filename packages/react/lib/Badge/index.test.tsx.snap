// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> should render 1`] = `
<div>
  <span
    class="junipero badge"
  >
    1
  </span>
</div>
`;
